pkg/server/commands/write_test.go

```diff
@@ -197,12 +209,227 @@ func TestTransactionalWriteFailedError(t *testing.T) {
-				tuple.NewTupleKey("document:1", "viewer", "user:jon"),
+		Writes: &openfgav1.WriteRequestTupleKeys{
+			TupleKeys: []*openfgav1.WriteRequestTupleKey{
+				{
+					Object:   "document:1",
+					Relation: "viewer",
+					User:     "user:jon",
+				},
 			},
 		},
 	})
-	require.ErrorIs(t, err, serverErrors.NewInternalError("concurrent write conflict", storage.ErrTransactionalWriteFailed))
+	require.ErrorIs(
+		t,
+		err,
+		serverErrors.NewInternalError(
+			"concurrent write conflict",
+			storage.ErrTransactionalWriteFailed,
+		),
+	)
 	require.Nil(t, resp)
 }
+
+func TestValidateConditionsInTuples(t *testing.T) {
+	type test struct {
+		name          string
+		tuple         *openfgav1.TupleKey
+		expectedError error
+	}
+
+	model := &openfgav1.AuthorizationModel{
+		Id:            ulid.Make().String(),
+		SchemaVersion: typesystem.SchemaVersion1_1,
+		TypeDefinitions: []*openfgav1.TypeDefinition{
+			{
+				Type: "user",
+			},
+			{
+				Type: "document",
+				Relations: map[string]*openfgav1.Userset{
+					"owner":  typesystem.This(),
+					"writer": typesystem.This(),
+					"viewer": typesystem.This(),
+				},
+				Metadata: &openfgav1.Metadata{
+					Relations: map[string]*openfgav1.RelationMetadata{
+						"owner": {
+							DirectlyRelatedUserTypes: []*openfgav1.RelationReference{
+								typesystem.DirectRelationReference("user", ""),
+								typesystem.ConditionedRelationReference(
+									typesystem.DirectRelationReference("user", ""),
+									"condition1",
+								),
+							},
+						},
+						"writer": {
+							DirectlyRelatedUserTypes: []*openfgav1.RelationReference{
+								typesystem.DirectRelationReference("user", ""),
+							},
+						},
+						"viewer": {
+							DirectlyRelatedUserTypes: []*openfgav1.RelationReference{
+								typesystem.ConditionedRelationReference(
+									typesystem.WildcardRelationReference("user"),
+									"condition1",
+								),
+							},
+						},
+					},
+				},
+			},
+		},
+		Conditions: map[string]*openfgav1.Condition{
+			"condition1": {
+				Name:       "condition1",
+				Expression: "param1 == 'ok'",
+				Parameters: map[string]*openfgav1.ConditionParamTypeRef{
+					"param1": {
+						TypeName: openfgav1.ConditionParamTypeRef_TYPE_NAME_STRING,
+					},
+				},
+			},
+		},
+	}
+
+	contextStructGood, err := structpb.NewStruct(map[string]interface{}{"param1": "ok"})
+	require.NoError(t, err)
+
+	contextStructBad, err := structpb.NewStruct(map[string]interface{}{"param1": "ok", "param2": 1})
+	require.NoError(t, err)
+
+	logger := logger.NewNoopLogger()
+
+	mockController := gomock.NewController(t)
+	defer mockController.Finish()
+
+	mockDatastore := mockstorage.NewMockOpenFGADatastore(mockController)
+	mockDatastore.EXPECT().MaxTuplesPerWrite().AnyTimes().Return(10)
+	mockDatastore.EXPECT().
+		ReadAuthorizationModel(gomock.Any(), gomock.Any(), gomock.Any()).
+		AnyTimes().
+		Return(model, nil)
+
+	cmd := NewWriteCommand(mockDatastore, logger)
+
+	tests := []test{
```

Comment:
Would you add a test to cover the case where you reference an undefined condition. For example, `condition2`.

If a tuple references a condition that is undefined in that model, then it should return an error.